Angela Diaz | 
ANA675 Neural Networks & Deep Learning |
Week 1 Assignment 1 |
7/13/2024

# Predict which digit an image represents

%cd "C:\Users\Angela\OneDrive\Desktop\ANA675"

In [118]:
# Import Packages
import pandas as pd
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split

In [119]:
# Load Dataset
digits = load_digits()

### Digit Data Exploration

In [120]:
# Set up dataframe
df = pd.DataFrame(digits.data, columns = digits.feature_names)

In [121]:
# How large is this dataset?
df.shape

(1797, 64)

In [122]:
# What does this dataset look like? Head
df.head()

pixel_0_0  pixel_0_1  pixel_0_2  pixel_0_3  pixel_0_4  pixel_0_5  \
0        0.0        0.0        5.0       13.0        9.0        1.0   
1        0.0        0.0        0.0       12.0       13.0        5.0   
2        0.0        0.0        0.0        4.0       15.0       12.0   
3        0.0        0.0        7.0       15.0       13.0        1.0   
4        0.0        0.0        0.0        1.0       11.0        0.0   

   pixel_0_6  pixel_0_7  pixel_1_0  pixel_1_1  ...  pixel_6_6  pixel_6_7  \
0        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1        0.0        0.0        0.0        0.0  ...        0.0        0.0   
2        0.0        0.0        0.0        0.0  ...        5.0        0.0   
3        0.0        0.0        0.0        8.0  ...        9.0        0.0   
4        0.0        0.0        0.0        0.0  ...        0.0        0.0   

   pixel_7_0  pixel_7_1  pixel_7_2  pixel_7_3  pixel_7_4  pixel_7_5  \
0        0.0        0.0        6.0       13.0       10.0        0.0   
1        0.0        0.0        0.0       11.0       16.0       10.0   
2        0.0        0.0        0.0        3.0       11.0       16.0   
3        0.0        0.0        7.0       13.0       13.0        9.0   
4        0.0        0.0        0.0        2.0       16.0        4.0   

   pixel_7_6  pixel_7_7  
0        0.0        0.0  
1        0.0        0.0  
2        9.0        0.0  
3        0.0        0.0  
4        0.0        0.0  

[5 rows x 64 columns]

In [123]:
# What does this dataset look like? Tail
df.tail()

pixel_0_0  pixel_0_1  pixel_0_2  pixel_0_3  pixel_0_4  pixel_0_5  \
1792        0.0        0.0        4.0       10.0       13.0        6.0   
1793        0.0        0.0        6.0       16.0       13.0       11.0   
1794        0.0        0.0        1.0       11.0       15.0        1.0   
1795        0.0        0.0        2.0       10.0        7.0        0.0   
1796        0.0        0.0       10.0       14.0        8.0        1.0   

      pixel_0_6  pixel_0_7  pixel_1_0  pixel_1_1  ...  pixel_6_6  pixel_6_7  \
1792        0.0        0.0        0.0        1.0  ...        4.0        0.0   
1793        1.0        0.0        0.0        0.0  ...        1.0        0.0   
1794        0.0        0.0        0.0        0.0  ...        0.0        0.0   
1795        0.0        0.0        0.0        0.0  ...        2.0        0.0   
1796        0.0        0.0        0.0        2.0  ...        8.0        0.0   

      pixel_7_0  pixel_7_1  pixel_7_2  pixel_7_3  pixel_7_4  pixel_7_5  \
1792        0.0        0.0        2.0       14.0       15.0        9.0   
1793        0.0        0.0        6.0       16.0       14.0        6.0   
1794        0.0        0.0        2.0        9.0       13.0        6.0   
1795        0.0        0.0        5.0       12.0       16.0       12.0   
1796        0.0        1.0        8.0       12.0       14.0       12.0   

      pixel_7_6  pixel_7_7  
1792        0.0        0.0  
1793        0.0        0.0  
1794        0.0        0.0  
1795        0.0        0.0  
1796        1.0        0.0  

[5 rows x 64 columns]

In [124]:
# What variables are in this dataset?
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1797 entries, 0 to 1796
Data columns (total 64 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   pixel_0_0  1797 non-null   float64
 1   pixel_0_1  1797 non-null   float64
 2   pixel_0_2  1797 non-null   float64
 3   pixel_0_3  1797 non-null   float64
 4   pixel_0_4  1797 non-null   float64
 5   pixel_0_5  1797 non-null   float64
 6   pixel_0_6  1797 non-null   float64
 7   pixel_0_7  1797 non-null   float64
 8   pixel_1_0  1797 non-null   float64
 9   pixel_1_1  1797 non-null   float64
 10  pixel_1_2  1797 non-null   float64
 11  pixel_1_3  1797 non-null   float64
 12  pixel_1_4  1797 non-null   float64
 13  pixel_1_5  1797 non-null   float64
 14  pixel_1_6  1797 non-null   float64
 15  pixel_1_7  1797 non-null   float64
 16  pixel_2_0  1797 non-null   float64
 17  pixel_2_1  1797 non-null   float64
 18  pixel_2_2  1797 non-null   float64
 19  pixel_2_3  1797 non-null   float64
 20  pixel_2_

In [125]:
# What are the general statistics for this dataset?
df.describe()

pixel_0_0    pixel_0_1    pixel_0_2    pixel_0_3    pixel_0_4  \
count     1797.0  1797.000000  1797.000000  1797.000000  1797.000000   
mean         0.0     0.303840     5.204786    11.835838    11.848080   
std          0.0     0.907192     4.754826     4.248842     4.287388   
min          0.0     0.000000     0.000000     0.000000     0.000000   
25%          0.0     0.000000     1.000000    10.000000    10.000000   
50%          0.0     0.000000     4.000000    13.000000    13.000000   
75%          0.0     0.000000     9.000000    15.000000    15.000000   
max          0.0     8.000000    16.000000    16.000000    16.000000   

         pixel_0_5    pixel_0_6    pixel_0_7    pixel_1_0    pixel_1_1  ...  \
count  1797.000000  1797.000000  1797.000000  1797.000000  1797.000000  ...   
mean      5.781859     1.362270     0.129661     0.005565     1.993879  ...   
std       5.666418     3.325775     1.037383     0.094222     3.196160  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
50%       4.000000     0.000000     0.000000     0.000000     0.000000  ...   
75%      11.000000     0.000000     0.000000     0.000000     3.000000  ...   
max      16.000000    16.000000    15.000000     2.000000    16.000000  ...   

         pixel_6_6    pixel_6_7    pixel_7_0    pixel_7_1    pixel_7_2  \
count  1797.000000  1797.000000  1797.000000  1797.000000  1797.000000   
mean      3.725097     0.206455     0.000556     0.279354     5.557596   
std       4.919406     0.984401     0.023590     0.934302     5.103019   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     1.000000   
50%       1.000000     0.000000     0.000000     0.000000     4.000000   
75%       7.000000     0.000000     0.000000     0.000000    10.000000   
max      16.000000    13.000000     1.000000     9.000000    16.000000   

         pixel_7_3    pixel_7_4    pixel_7_5    pixel_7_6    pixel_7_7  
count  1797.000000  1797.000000  1797.000000  1797.000000  1797.000000  
mean     12.089037    11.809126     6.764051     2.067891     0.364496  
std       4.374694     4.933947     5.900623     4.090548     1.860122  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%      11.000000    10.000000     0.000000     0.000000     0.000000  
50%      13.000000    14.000000     6.000000     0.000000     0.000000  
75%      16.000000    16.000000    12.000000     2.000000     0.000000  
max      16.000000    16.000000    16.000000    16.000000    16.000000  

[8 rows x 64 columns]

### Digits Modeling

In [126]:
# Split data between training (80%) and testing (20%) sets 
X= digits.data
y = digits.target

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2)

#Check the length of training dataset
len(X_train)

1437

#### SVC

In [129]:
# Import library for SVC
from sklearn.svm import SVC

In [142]:
# Create Model
svc_model1 = SVC()
svc_model1.fit(X_train, y_train)

# Make Predictions
svc_model1.predict(X_test)

# Evaluate SVC Model
svc_model1.score(X_test, y_test)

0.9694444444444444

In [147]:
# Create Model
svc_model2 = SVC(C=3)
svc_model2.fit(X_train, y_train)

# Make Predictions
svc_model2.predict(X_test)

# Evaluate SVC Model
svc_model2.score(X_test, y_test)

0.9777777777777777

In [148]:
# Create Model
svc_model = SVC(C=5)
svc_model.fit(X_train, y_train)

SVC(C=5)

In [149]:
# Make Predictions
svc_model.predict(X_test) 

array([1, 9, 0, 9, 2, 6, 1, 3, 3, 3, 3, 3, 6, 9, 8, 9, 3, 6, 7, 3, 4, 2,
       0, 7, 5, 0, 7, 0, 9, 6, 0, 5, 3, 9, 4, 6, 0, 2, 4, 8, 6, 3, 2, 3,
       1, 9, 0, 6, 7, 4, 6, 0, 5, 8, 1, 1, 2, 0, 0, 9, 3, 5, 4, 6, 9, 7,
       0, 9, 7, 1, 1, 3, 0, 3, 2, 6, 3, 4, 0, 9, 8, 7, 4, 2, 9, 9, 7, 6,
       9, 4, 8, 4, 3, 7, 3, 5, 6, 3, 5, 0, 1, 7, 2, 1, 9, 7, 7, 8, 3, 9,
       6, 7, 9, 5, 3, 1, 8, 4, 8, 6, 4, 1, 5, 9, 8, 6, 6, 5, 9, 6, 7, 0,
       1, 9, 6, 8, 0, 3, 1, 0, 4, 2, 0, 5, 5, 5, 1, 7, 2, 8, 0, 2, 7, 2,
       7, 4, 0, 2, 5, 1, 0, 4, 8, 0, 5, 7, 0, 1, 1, 6, 4, 8, 8, 0, 1, 6,
       5, 3, 3, 6, 9, 2, 4, 1, 6, 8, 3, 2, 8, 5, 9, 7, 4, 4, 8, 3, 3, 4,
       6, 8, 9, 5, 3, 4, 7, 5, 8, 5, 1, 3, 8, 7, 2, 7, 8, 3, 6, 0, 3, 8,
       9, 5, 1, 5, 4, 2, 2, 5, 9, 2, 4, 3, 2, 9, 2, 6, 0, 3, 4, 7, 7, 5,
       2, 6, 8, 9, 9, 6, 2, 6, 7, 6, 4, 5, 2, 9, 4, 7, 6, 8, 5, 8, 1, 0,
       2, 2, 9, 2, 7, 6, 4, 9, 3, 6, 7, 9, 4, 6, 4, 4, 0, 4, 9, 4, 3, 8,
       6, 2, 2, 2, 7, 9, 8, 8, 5, 5, 8, 8, 1, 6, 1,

In [150]:
# Evaluate SVC Model
svc_model.score(X_test, y_test)

0.9805555555555555

#### SVC with Hypertuning

In [151]:
# Hypertuning
from sklearn.model_selection import GridSearchCV 

param_grid = {'C': [0.1, 1, 10, 100, 1000],  
              'gamma': [1, 0.1, 0.01, 0.001, 0.0001], 
              'kernel': ['rbf']}  
  
grid = GridSearchCV(SVC(), param_grid, refit = True, verbose = 3) 
grid.fit(X_train, y_train) 

Fitting 5 folds for each of 25 candidates, totalling 125 fits
[CV 1/5] END ........C=0.1, gamma=1, kernel=rbf;, score=0.108 total time=   0.1s
[CV 2/5] END ........C=0.1, gamma=1, kernel=rbf;, score=0.108 total time=   0.1s
[CV 3/5] END ........C=0.1, gamma=1, kernel=rbf;, score=0.108 total time=   0.1s
[CV 4/5] END ........C=0.1, gamma=1, kernel=rbf;, score=0.105 total time=   0.1s
[CV 5/5] END ........C=0.1, gamma=1, kernel=rbf;, score=0.108 total time=   0.1s
[CV 1/5] END ......C=0.1, gamma=0.1, kernel=rbf;, score=0.108 total time=   0.0s
[CV 2/5] END ......C=0.1, gamma=0.1, kernel=rbf;, score=0.108 total time=   0.1s
[CV 3/5] END ......C=0.1, gamma=0.1, kernel=rbf;, score=0.108 total time=   0.0s
[CV 4/5] END ......C=0.1, gamma=0.1, kernel=rbf;, score=0.105 total time=   0.0s
[CV 5/5] END ......C=0.1, gamma=0.1, kernel=rbf;, score=0.108 total time=   0.0s
[CV 1/5] END .....C=0.1, gamma=0.01, kernel=rbf;, score=0.108 total time=   0.0s
[CV 2/5] END .....C=0.1, gamma=0.01, kernel=rbf

[CV 2/5] END .......C=1000, gamma=1, kernel=rbf;, score=0.108 total time=   0.1s
[CV 3/5] END .......C=1000, gamma=1, kernel=rbf;, score=0.108 total time=   0.1s
[CV 4/5] END .......C=1000, gamma=1, kernel=rbf;, score=0.105 total time=   0.1s
[CV 5/5] END .......C=1000, gamma=1, kernel=rbf;, score=0.108 total time=   0.1s
[CV 1/5] END .....C=1000, gamma=0.1, kernel=rbf;, score=0.108 total time=   0.0s
[CV 2/5] END .....C=1000, gamma=0.1, kernel=rbf;, score=0.108 total time=   0.1s
[CV 3/5] END .....C=1000, gamma=0.1, kernel=rbf;, score=0.108 total time=   0.0s
[CV 4/5] END .....C=1000, gamma=0.1, kernel=rbf;, score=0.105 total time=   0.0s
[CV 5/5] END .....C=1000, gamma=0.1, kernel=rbf;, score=0.108 total time=   0.1s
[CV 1/5] END ....C=1000, gamma=0.01, kernel=rbf;, score=0.858 total time=   0.0s
[CV 2/5] END ....C=1000, gamma=0.01, kernel=rbf;, score=0.844 total time=   0.0s
[CV 3/5] END ....C=1000, gamma=0.01, kernel=rbf;, score=0.819 total time=   0.0s
[CV 4/5] END ....C=1000, gam

GridSearchCV(estimator=SVC(),
             param_grid={'C': [0.1, 1, 10, 100, 1000],
                         'gamma': [1, 0.1, 0.01, 0.001, 0.0001],
                         'kernel': ['rbf']},
             verbose=3)

In [152]:
# Make Predictions
grid.predict(X_test) 

array([1, 9, 0, 9, 2, 6, 1, 3, 3, 3, 3, 3, 6, 9, 8, 9, 3, 6, 7, 3, 4, 2,
       0, 7, 5, 0, 7, 0, 9, 6, 0, 5, 3, 9, 4, 6, 0, 2, 4, 8, 6, 3, 2, 3,
       1, 9, 0, 6, 7, 4, 6, 0, 5, 8, 1, 1, 2, 0, 0, 9, 3, 5, 4, 6, 9, 7,
       0, 9, 7, 1, 1, 3, 0, 3, 2, 6, 3, 4, 0, 9, 8, 7, 4, 2, 9, 9, 7, 6,
       9, 4, 8, 4, 3, 7, 3, 5, 6, 3, 5, 0, 1, 7, 2, 1, 9, 7, 7, 8, 3, 9,
       6, 7, 9, 5, 3, 1, 8, 4, 8, 6, 4, 1, 5, 9, 8, 6, 6, 5, 9, 6, 7, 0,
       1, 9, 6, 8, 0, 3, 1, 0, 4, 2, 0, 5, 5, 5, 1, 7, 2, 8, 0, 2, 7, 2,
       7, 4, 0, 2, 5, 1, 0, 4, 8, 0, 5, 7, 0, 1, 1, 6, 4, 8, 8, 0, 1, 6,
       5, 3, 3, 6, 9, 2, 4, 1, 6, 8, 3, 2, 3, 5, 9, 7, 4, 4, 8, 3, 3, 4,
       6, 8, 9, 5, 3, 4, 7, 5, 8, 5, 1, 3, 8, 7, 2, 7, 8, 3, 6, 0, 3, 8,
       9, 5, 1, 5, 4, 2, 2, 5, 9, 2, 4, 3, 2, 9, 3, 6, 0, 3, 4, 7, 7, 5,
       2, 6, 8, 9, 9, 6, 2, 6, 7, 6, 4, 5, 2, 9, 4, 7, 6, 8, 5, 8, 1, 0,
       2, 2, 9, 2, 7, 6, 4, 9, 3, 6, 7, 9, 4, 6, 4, 4, 0, 4, 9, 4, 3, 8,
       6, 2, 2, 2, 7, 9, 8, 8, 5, 5, 8, 8, 1, 6, 1,

In [153]:
# Evaluate grid Model
grid.score(X_test, y_test)

0.9833333333333333

#### GaussianNB

In [154]:
# Import library for GaussuanNB
from sklearn.naive_bayes import GaussianNB

In [155]:
# Create Model
gnb = GaussianNB()
gnb.fit(X_train, y_train)

GaussianNB()

In [156]:
# Make Predictions
gnb.predict(X_test) 

array([1, 1, 4, 9, 2, 6, 1, 3, 3, 3, 3, 3, 6, 9, 8, 9, 3, 6, 7, 8, 4, 2,
       0, 7, 5, 0, 7, 0, 9, 6, 0, 5, 3, 3, 4, 6, 0, 2, 4, 7, 6, 3, 8, 3,
       1, 9, 0, 6, 7, 4, 6, 0, 5, 8, 1, 1, 2, 0, 0, 9, 3, 5, 4, 6, 9, 7,
       0, 7, 7, 1, 6, 3, 0, 3, 2, 6, 3, 4, 0, 9, 8, 7, 4, 2, 8, 9, 7, 6,
       7, 4, 8, 7, 7, 7, 3, 5, 6, 8, 5, 0, 1, 7, 8, 1, 9, 7, 7, 7, 3, 9,
       4, 7, 4, 5, 3, 1, 8, 4, 8, 6, 4, 1, 5, 9, 8, 6, 6, 5, 9, 6, 7, 0,
       1, 9, 6, 5, 0, 3, 1, 0, 4, 1, 0, 5, 5, 5, 1, 7, 3, 8, 0, 3, 7, 2,
       7, 4, 0, 2, 5, 1, 0, 4, 8, 0, 5, 7, 0, 1, 1, 6, 4, 8, 8, 0, 1, 6,
       5, 3, 3, 6, 3, 8, 4, 1, 6, 8, 3, 8, 3, 5, 9, 7, 4, 4, 8, 3, 3, 4,
       6, 8, 7, 5, 3, 4, 7, 5, 8, 7, 1, 3, 8, 7, 8, 7, 8, 3, 6, 0, 3, 8,
       3, 7, 1, 5, 4, 2, 2, 5, 9, 2, 4, 3, 8, 9, 8, 6, 0, 3, 4, 7, 7, 7,
       8, 6, 8, 9, 3, 6, 2, 6, 7, 6, 4, 5, 8, 8, 4, 7, 6, 8, 5, 7, 6, 0,
       2, 3, 3, 2, 7, 6, 4, 9, 3, 6, 7, 9, 4, 6, 4, 4, 0, 7, 9, 4, 3, 8,
       6, 2, 2, 8, 7, 3, 8, 8, 5, 5, 8, 8, 1, 5, 1,

In [157]:
# Evaluate GaussianNB Model
gnb.score(X_test, y_test)

0.8388888888888889

#### KNeighbors Classifier

In [164]:
# Import Library for KNeighbors Classifier
from sklearn.neighbors import KNeighborsClassifier

In [171]:
# Create Model
knc1 = KNeighborsClassifier(n_neighbors = 10)
knc1.fit(X_train, y_train)

# Make Predictions
knc1.predict(X_test) 

# Evaluate KNeighborsClassifier Model
knc1.score(X_test, y_test)

0.9722222222222222

In [176]:
# Create Model
knc2 = KNeighborsClassifier(n_neighbors = 8)
knc2.fit(X_train, y_train)

# Make Predictions
knc2.predict(X_test) 

# Evaluate KNeighborsClassifier Model
knc2.score(X_test, y_test)

0.975

In [177]:
# Create Model
knc = KNeighborsClassifier(n_neighbors = 5)
knc.fit(X_train, y_train)

KNeighborsClassifier()

In [178]:
# Make Predictions
knc.predict(X_test) 

array([1, 1, 0, 9, 2, 6, 1, 3, 3, 3, 3, 3, 6, 9, 8, 9, 3, 6, 7, 3, 4, 2,
       0, 7, 5, 0, 7, 0, 9, 6, 0, 5, 3, 9, 4, 6, 0, 2, 4, 8, 6, 3, 2, 3,
       1, 9, 0, 6, 7, 4, 6, 0, 5, 8, 1, 1, 2, 0, 0, 9, 3, 5, 4, 6, 9, 7,
       0, 7, 7, 1, 1, 3, 0, 3, 2, 6, 3, 4, 0, 9, 8, 7, 4, 2, 9, 9, 7, 6,
       9, 4, 8, 4, 3, 7, 3, 5, 6, 8, 5, 0, 1, 7, 2, 1, 9, 7, 7, 8, 3, 9,
       6, 7, 9, 5, 3, 1, 8, 4, 8, 6, 4, 1, 5, 9, 8, 6, 6, 5, 9, 6, 7, 0,
       1, 9, 6, 8, 0, 3, 1, 0, 4, 2, 0, 5, 5, 5, 1, 7, 2, 8, 0, 2, 7, 2,
       7, 4, 0, 2, 5, 1, 0, 4, 8, 0, 5, 7, 0, 1, 1, 6, 4, 8, 8, 0, 1, 6,
       5, 3, 3, 6, 9, 2, 4, 1, 6, 8, 3, 2, 3, 5, 9, 7, 4, 4, 8, 3, 3, 4,
       6, 8, 9, 5, 3, 4, 7, 5, 8, 5, 1, 3, 8, 7, 2, 7, 8, 3, 6, 0, 3, 8,
       9, 5, 1, 5, 4, 2, 2, 5, 9, 2, 4, 3, 2, 9, 3, 6, 0, 3, 4, 7, 7, 5,
       2, 6, 8, 9, 9, 6, 2, 6, 7, 6, 4, 5, 2, 9, 4, 7, 6, 8, 5, 8, 1, 0,
       2, 2, 9, 2, 7, 6, 4, 9, 3, 6, 7, 9, 4, 6, 4, 4, 0, 4, 9, 4, 3, 8,
       6, 2, 2, 2, 7, 9, 8, 8, 5, 5, 8, 8, 1, 6, 1,

In [179]:
# Evaluate KNeighborsClassifier Model
knc.score(X_test, y_test)

0.9777777777777777

### Best Model(s)

GridSearchCV is the best model out of all models with an accuracy score of 0.9833333333333333. The second best model is the regular SVC model which has an accuracy of 0.9805555555555555. The third best model is KNeighbors Classifier with an accuracy score of 0.9777777777777777. GaussianNB is the least best model which has an accuracy score of 0.8388888888888889. 